src/systems/detachable_joint/DetachableJoint.cc

```diff
@@ -234,8 +234,75 @@ void DetachableJoint::PreUpdate(
     }
     else
     {
-      modelEntity = _ecm.EntityByComponents(
-          components::Model(), components::Name(this->childModelName));
+      auto entitiesMatchingName = entitiesFromScopedName(
+        this->childModelName, _ecm);
+
+      // TODO(arjoc): There is probably a more efficient way of combining entitiesFromScopedName
+      // With filtering.
+      // Filter for entities with only models
+      std::vector<Entity> candidateEntities;
+      for (const auto entity: entitiesMatchingName)
+      {
+        if (_ecm.EntityHasComponentType(entity, components::Model::typeId))
+        {
+          candidateEntities.push_back(entity);
+        }
+        else
+        {
+          ignwarn << "Found entity " << scopedName(entity, _ecm)
+            << " but it is not a model." << std::endl;
+        }
+      }
+
+      if (candidateEntities.size() == 1)
+      {
+        // If there is one entity select that entity itself
+        modelEntity = *candidateEntities.begin();
+      }
+      else
+      {
+        std::string selectedModelName;
+        auto parentEntityScopedPath = scopedName(this->model.Entity(), _ecm);
+        // If there is more than one entity with the same name, look for the
+        // entity with a parent component.
+        for (auto entity : candidateEntities)
+        {
+          auto childEntityScope = scopedName(entity, _ecm);
+          if (childEntityScope.size() < parentEntityScopedPath.size())
+          {
+            continue;
+          }
+          if (childEntityScope.find(parentEntityScopedPath) != 0)
+          {
+            continue;
+          }
+          if (modelEntity == kNullEntity)
+          {
+
+            this->childLinkEntity = _ecm.EntityByComponents(
+                      components::Link(), components::ParentEntity(entity),
+                      components::Name(this->childLinkName));
+
+            if (kNullEntity != this->childLinkEntity)
+            {
+                  // Only select child entity if tbe entity has a link
+                  modelEntity = entity;
+                  selectedModelName = childEntityScope;
+                  igndbg << "Selecting " << childEntityScope << " as model to be detached" << std::endl;
+            }
+            else
+            {
+              ignwarn << "Found " << childEntityScope << " with no link named " << this->childLinkName <<std::endl;
+              this->suppressChildWarning = true;
```

Comment:
I don't think we should set `this->suppressChildWarning` as it seems like an input parameter for the plugin? See line 208

Reply:
[1a7083d](https://github.com/gazebosim/gz-sim/pull/1097/commits/1a7083dfeb05bd50e6358bbbd3377f6a13e0bb4e)